### PR TITLE
Honor `created()` function for an item in a submenu.

### LINKED ISSUE #10483: The created function is not honored for a menu item in a submenu
I believe it is due to this code:

https://github.com/atom/atom/blob/v1.4.0/src/context-menu-manager.coffee#L140-L144

Specifically, `Object.create()` is called on a root `item`, is screened for visibility via `shouldDisplay()`, and then its `created()` function is called, if it exists.

Unfortunately, this sequence is not applied recursively for submenu items.

### PR DESCRIPTION
**Pull request by bolinfest**
Fixes #10483.

### CODE REVIEW

**Comment by bolinfest** [comment]
I'm curious if there's any chance of getting this patched in on the v1.5.0 release branch?

File: src/context-menu-manager.coffee [hunk 1]
```diff
@@ -150,6 +147,19 @@ class ContextMenuManager
...
+  # Returns an object compatible with `::add()` or `null`.
+  cloneItemForEvent: (item, event) ->
+    return null if item.devMode and not @devMode
+    item = Object.create(item)
+    if typeof item.shouldDisplay is 'function'
+      return null unless item.shouldDisplay(event)
+    item.created?(event)
+    if Array.isArray(item.submenu)
+      item.submenu = item.submenu
+        .map((item) => @cloneItemForEvent(item, event))
```
Comment (ssorallen) [inline src/context-menu-manager.coffee]: These `item` variables shadow the outer `item`, which made this a bit tough to read the first time.

Comment (bolinfest) [inline src/context-menu-manager.coffee]: OK, changed.

**Comment by bolinfest** [comment]
@nathansobo @maxbrunsfeld I'm hoping this should not be contentious since I believe it fixes a bug where the implementation does not match the spec: would either of you be able to take a look?

I'm not sure if there are CoffeeScript style changes or extra comments you'd like to see, but let me know and I'll make it happen. Thanks guys!

**Comment by bolinfest** [comment]
I'd really like to get this in before the next branch is cut because this is blocking some menu management improvements I'm trying to make to Nuclide. Still trying to find a reviewer: @joshaber @mnquintana would you be able to take a look?

**Comment by maxbrunsfeld** [comment]
@bolinfest Last week was the Atom team summit, but we will be able to take a look at this on Monday.

**Comment by bolinfest** [comment]
@maxbrunsfeld Ah, sorry, didn't realize: thanks and have a good weekend!

File: spec/context-menu-manager-spec.coffee [hunk 2]
```diff
@@ -156,3 +156,28 @@ describe "ContextMenuManager", ->
       catch error
         addError = error
       expect(addError.message).toContain('<>')
+
+    it "recursively applies templateForEvent on submenu items", ->
```
Comment (maxbrunsfeld) [inline spec/context-menu-manager-spec.coffee]: This description is a little bit implementation-specific. How about something like

```coffee
it "calls `created` hooks for submenu items", ->
```

**Comment by maxbrunsfeld** [comment]
I left one small suggestion about the test. Otherwise, this looks great. Thanks for the fix!

**Comment by bolinfest** [comment]
@maxbrunsfeld Thanks: done!

**Pull request merged**